Add `SetEntryAssembly()` API to `System.Reflection`

ivdiazsa opened the pull request:
Implements the API proposal detailed and approved in issue #101616. This new API will allow developers to change the entry assembly of their .NET apps on the fly, should they require it. One important scenario is app launchers. With the usage of this API, then functions like `GetEntryAssembly()` will return the right value, and thus we will be able to ensure the information is consistent and correct.

ivdiazsa committed:
Trying out JanK's approach to override the entry assembly...

ivdiazsa committed:
Fixed what was missing for this reflection scenario to work correctly.

kg commented on src/coreclr/System.Private.CoreLib/CompatibilitySuppressions.xml:
```diff
@@ -1,6 +1,22 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <!-- https://learn.microsoft.com/en-us/dotnet/fundamentals/package-validation/diagnostic-ids -->
 <Suppressions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
+  <Suppression>
```
Why are these new suppressions needed?

ivdiazsa replied:
Oh that was done automatically by the build script, so I don't know. We could try reverting it and generating it again if one or more of these are confirmed to be unnecessary

jkotas replied:
You have more than one month old sync of the main branch. Try rebasing against the current main.

It is not expected that the build scripts do edits like this.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs:
```diff
@@ -218,12 +218,17 @@ public override string ToString()
-        private static bool s_forceNullEntryPoint;
+        private static Assembly? s_overriddenEntryAssembly;
+
+        public static void SetEntryAssembly(Assembly newAssembly)
```
The approved argument name is `assembly`.

Also, I think this should allow passing in `null` to clear the entry assembly.

ivdiazsa replied:
Is there a reason or use case we would want to clear the entry assembly entirely? I see it from the perspective that a dotnet app will always have an entry assembly, otherwise how is it executing, so I think clearing the entry assembly would just be inconsistent with the program's state.

MichalStrehovsky replied:
It's null for various hosting scenarios or e.g. native libraries: https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/libraries

jkotas replied:
> I see it from the perspective that a dotnet app will always have an entry assembly

The entry assembly can be null in hosted scenarios when you start the runtime via hosting APIs (https://learn.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting) and not by launching an app.

Notice that the existing `Assembly.GetEntryAssembly` returns `Assembly?`.

> Is there a reason or use case we would want to clear the entry assembly entirely? 

One use case are tests that exercise the hosting scenario, without actually hosting the runtime. You can see we have number of those in this repo.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs:
```diff
@@ -218,12 +218,17 @@ public override string ToString()
 
         // internal test hook
-        private static bool s_forceNullEntryPoint;
+        private static Assembly? s_overriddenEntryAssembly;
```
You will need to replace `s_forceNullEntryPoint` uses in the tests with calls of this API.

https://github.com/search?q=repo%3Adotnet%2Fruntime%20s_forceNullEntryPoint&type=code

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs:
```diff
@@ -218,12 +218,17 @@ public override string ToString()
-        private static bool s_forceNullEntryPoint;
+        private static Assembly? s_overriddenEntryAssembly;
+
+        public static void SetEntryAssembly(Assembly newAssembly)
```
This needs API doc comment.

MichalStrehovsky commented on src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs:
```diff
@@ -218,12 +218,17 @@ public override string ToString()
+
+        public static void SetEntryAssembly(Assembly newAssembly)
+        {
+            s_overriddenEntryAssembly = newAssembly;
```
Do we need any validation. We discussed at least validating it's a runtime-defined assembly: https://github.com/dotnet/runtime/issues/101616#issuecomment-2096031589

ivdiazsa committed:
Merge branch 'main' into managed-contract-facade

ivdiazsa committed:
Reverted the compatibility suppressions added by the build script.

ivdiazsa committed:
Forgot to revert also the nativeaot part of the suppressions.

ivdiazsa committed:
Addressed PR comments: Updated the tests to use the new
"s_overriddenEntryAssembly", added an API doc to SetEntryAssembly(),
added validation for it to be a Runtime Assembly, and changed the type
to allow a null entry assembly.

jkotas commented on src/libraries/System.Configuration.ConfigurationManager/tests/System/Configuration/CustomHostTests.cs:
```diff
@@ -36,7 +36,7 @@ public void FilePathIsPopulatedCorrectly()
         private static void MakeAssemblyGetEntryAssemblyReturnNull()
         {
             typeof(Assembly)
```
This should call `Assembly.SetEntryAssembly(null)` instead of private reflection.

ivdiazsa replied:
I was actually wondering that something didn't look quite right there, but couldn't put my finger on it. Will fix

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs:
```diff
@@ -217,13 +217,31 @@ public override string ToString()
+        /// </param>
+        /// <remarks>
+        /// It is important to mention that the assembly passed to this function
+        /// has to be a RuntimeAssembly type. Otherwise, an exception will be thrown.
```
`RuntimeAssembly` type is internal implementation detail of the runtime. We do not document what it is.

This should say "a runtime Assembly object" or "runtime defined Assembly object".

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs:
```diff
@@ -217,13 +217,31 @@ public override string ToString()
+            if (assembly is not RuntimeAssembly)
+                throw new ArgumentException(SR.Argument_MustBeRuntimeAssembly);
+
+            s_overriddenEntryAssembly = assembly;
```
If you call `Assembly.SetEntryAssembly(null)`, it should make GetEntryAssembly to return null. (The current implementation makes it return the default original value.)

Here one possible way how it can done that I have shared earlier: https://github.com/jkotas/runtime/commit/a6a0246212fc9548d928e8b7f2459c8ccbd32937 (there are certainly other options too)

jkotas commented:
This need some tests ( https://github.com/dotnet/runtime/blob/1ee61b8667d9db31c0fe53d2e1abc274fb486cfc/src/libraries/System.Runtime/tests/System.Reflection.Tests/AssemblyTests.cs#L182 is a good place to add them)

ivdiazsa committed:
Added tests and addressed more comments on the PR.

jkotas commented on src/libraries/System.Runtime/tests/System.Reflection.Tests/AssemblyTests.cs:
```diff
@@ -181,6 +181,19 @@ public void GetEntryAssembly()
         }
 
+        [Fact]
+        public void SetEntryAssembly()
```
This test may want to use `RemoteExecutor.Invoke(` to avoid interfering with the existing GetEntryAssembly test.

jkotas commented on src/libraries/System.Runtime/tests/System.Reflection.Tests/AssemblyTests.cs:
```diff
@@ -181,6 +181,19 @@ public void GetEntryAssembly()
+            Assembly.SetEntryAssembly(null);
+            Assert.Null(Assembly.GetEntryAssembly());
+
+            Assembly.SetEntryAssembly(originalAssembly);
```
```suggestion
            Assembly.SetEntryAssembly(typeof(AssemblyTests).Assembly);
```
This may be more interesting test case.

jkotas commented on src/libraries/System.Runtime/tests/System.Reflection.Tests/AssemblyTests.cs:
```diff
@@ -181,6 +181,19 @@ public void GetEntryAssembly()
+
+            Assembly.SetEntryAssembly(originalAssembly);
+            Assert.Equal(Assembly.GetEntryAssembly(), originalAssembly);
+        }
```
Add a test case that verifies we throw for non-runtime assembly. You can create one of those e.g. by calling `new PersistedAssemblyBuilder(...)`.

ivdiazsa committed:
Added exception test case for SetEntryAssembly, and wrapped all its
test cases in a RemoteExecutor.Invoke() call, in order to avoid
potentially interferring with the GetEntryAssembly tests.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs:
```diff
@@ -217,13 +217,38 @@ public override string ToString()
+        {
+            if (assembly is null)
+            {
+                s_overriddenEntryAssembly = new object();
```
```suggestion
                s_overriddenEntryAssembly = string.Empty;
```
This is a tiny bit more efficient placeholder.

ivdiazsa committed:
Update src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

ivdiazsa commented:
Seems I got an import wrong. Will look into it.

jkotas commented on src/libraries/System.Configuration.ConfigurationManager/tests/System/Configuration/CustomHostTests.cs:
```diff
@@ -31,14 +31,12 @@ public void FilePathIsPopulatedCorrectly()
+        /// Makes Assembly.GetEntryAssembly() return null by passing the null literal
+        /// to Assembly.SetEntryAssembly().
         /// </summary>
         private static void MakeAssemblyGetEntryAssemblyReturnNull()
```
I think you can delete this method and just call `Assembly.SetEntryAssembly(null);` in the one place that it is used. (The factoring into separate method made sense when the method was more complicated.)

jkotas commented on src/libraries/System.Diagnostics.TraceSource/tests/System.Diagnostics.TraceSource.Tests/DefaultTraceListenerClassTests.cs:
```diff
@@ -169,14 +169,12 @@ public void EntryAssemblyName_Null_NotIncludedInTrace()
+        /// Makes Assembly.GetEntryAssembly() return null by passing the null literal
+        /// to Assembly.SetEntryAssembly().
         /// </summary>
         private static void MakeAssemblyGetEntryAssemblyReturnNull()
```
Same here

ivdiazsa committed:
Refactored further the tests that force a null entry assembly, and
fixed the issue with the remote executor.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs:
```diff
@@ -217,13 +217,38 @@ public override string ToString()
+        private static object? s_overriddenEntryAssembly;
+
+        /// <summary>
+        /// Sets the application's entry assembly to the provided assembly object
+        /// as argument.
```
```suggestion
        /// Sets the application's entry assembly to the provided assembly object.
```

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs:
```diff
@@ -217,13 +217,38 @@ public override string ToString()
+        /// Assembly object that represents the application's new entry assembly.
+        /// </param>
+        /// <remarks>
+        /// It is important to mention that the assembly passed to this function
+        /// has to be a runtime defined Assembly type object. Otherwise, an exception
+        /// will be thrown.
```
```suggestion
        /// The assembly passed to this function has to be a runtime defined Assembly 
        /// type object. Otherwise, an exception will be thrown.
```
"It is important to mention that ... " is not a style used in our official docs.

ivdiazsa replied:
Ok noted for future docs I might write.

jkotas commented on src/libraries/System.Runtime/tests/System.Reflection.Tests/AssemblyTests.cs:
```diff
@@ -181,6 +183,32 @@ public void GetEntryAssembly()
+                Assembly.SetEntryAssembly(null);
+                Assert.Null(Assembly.GetEntryAssembly());
+
+                Assembly originalAssembly = typeof(AssemblyTests).Assembly;
+
+                Assembly.SetEntryAssembly(originalAssembly);
+                Assert.Equal(Assembly.GetEntryAssembly(), originalAssembly);
```
```suggestion
                Assembly testAssembly = typeof(AssemblyTests).Assembly;

                Assembly.SetEntryAssembly(testAssembly testAssembly);
                Assert.Equal(Assembly.GetEntryAssembly(), testAssembly);
```
This is not the original assembly

ivdiazsa replied:
I'm confused. If this is not the original assembly, then which is? Can you explain more in detail about this change?

jkotas replied:
I would expect that the original value of `Assembly.GetEntryAssembly()` to be `Microsoft.DotNet.RemoteExecutor`, not the test assembly.

ivdiazsa replied:
I think you're right, that makes sense. I forgot that we were now running within the remote executor. Thanks for clarifying Jan

jkotas commented on src/libraries/System.Runtime/tests/System.Reflection.Tests/AssemblyTests.cs:
```diff
@@ -181,6 +183,32 @@ public void GetEntryAssembly()
             Assert.True(correct, $"Unexpected assembly name {assembly}");
         }
 
+        [Fact]
```
```suggestion
        [ConditionalTheory(typeof(RemoteExecutor), nameof(RemoteExecutor.IsSupported))]
```
to fix test failures

ivdiazsa committed:
Apply Jan's suggestions

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

jkotas reviewed (approved):
LGTM. Thank you!

ivdiazsa commented:
I wonder where that trailing whitespace came from... Let me fix it so we can merge. Thanks a lot for your guidance with this @jkotas!

ivdiazsa committed:
Fixed a sneaky trailing whitespace that was messing up with the code analyzers.

ivdiazsa commented:
Might be just my ptsd of breaking the outerloop with a PR I did last year, so I'll kick off an outerloop run just to be safe we can merge this.

ivdiazsa commented:
/azp run runtime-coreclr outerloop

ivdiazsa committed:
Changed ConditionalTheory to ConditionalFact in the tests because we're not passing parameters to it.

ivdiazsa committed:
Disabled building the CustomHostTests test file when .NET Framework
